src/Xrefcheck/Verify.hs

```diff
@@ -81,230 +87,324 @@ toVerifyRes = VerifyResult . either one (\() -> [])
+
+checkExternalResource :: VerifyConfig -> Text -> IO (VerifyResult VerifyError)
 checkExternalResource VerifyConfig{..} link
-    | isIgnored = return mempty
-    | doesReferLocalhost = return mempty
-    | otherwise = fmap toVerifyRes $ do
-        makeRequest HEAD 0.3 >>= \case
-            Right () -> return $ Right ()
-            Left   _ -> makeRequest GET 0.7
+  | isIgnored = return mempty
+  | doesReferLocalhost = return mempty
+  | otherwise = fmap toVerifyRes $ runExceptT $ do
+      uri <- mkURI link & maybe (throwError ExternalResourceInvalidUri) pure
+
+      case unRText <$> uriScheme uri of
+        Just "http" -> checkHttp uri
+        Just "https" -> checkHttp uri
+        Just "ftp" -> checkFtp uri False
+        Just "ftps" -> checkFtp uri True
+        _ -> throwError ExternalResourceUnknownProtocol
   where
     isIgnored =
-        let maybeIsIgnored = (doesMatchAnyRegex link) <$> vcIgnoreRefs
-        in fromMaybe False maybeIsIgnored
-    doesReferLocalhost = any (`T.isInfixOf` link) ["://localhost", "://127.0.0.1"]
+      let maybeIsIgnored = doesMatchAnyRegex link <$> vcIgnoreRefs
+      in fromMaybe False maybeIsIgnored
+
+    doesReferLocalhost = any (`T.isInfixOf` link) $
+      ["://localhost", "://127.0.0.1"]
 
     doesMatchAnyRegex :: Text -> ([Regex] -> Bool)
     doesMatchAnyRegex src = any $ \regex ->
-        case regexec regex src of
-            Right res -> case res of
-                Just (before, match, after, _) -> null before && null after && not (null match)
-                Nothing -> False
-            Left _ -> False
-
-    makeRequest :: _ => method -> RatioNat -> IO (Either VerifyError ())
-    makeRequest method timeoutFrac = runExceptT $ do
-        uri <- mkURI link
-             & maybe (throwError ExternalResourceInvalidUri) pure
-        parsedUrl <- useURI uri
-                   & maybe (throwError ExternalResourceUnknownProtocol) pure
-        let reqLink = case parsedUrl of
-                Left (url, option) ->
-                    runReq defaultHttpConfig $
-                    req method url NoReqBody ignoreResponse option
-                Right (url, option) ->
-                    runReq defaultHttpConfig $
-                    req method url NoReqBody ignoreResponse option
-
-        let maxTime = Time @Second $ unTime vcExternalRefCheckTimeout * timeoutFrac
-
-        mres <- liftIO (timeout maxTime $ void reqLink)
-                `catch` (either throwError (\() -> return (Just ())) . interpretErrors)
-        maybe (throwError $ ExternalResourceSomeError "Response timeout") pure mres
+      case regexec regex src of
+        Right res -> case res of
+          Just (before, match, after, _) ->
+            null before && null after && not (null match)
+          Nothing -> False
+        Left _ -> False
+
+    checkHttp :: URI -> ExceptT VerifyError IO ()
+    checkHttp uri = do
+      res <- liftIO $ runExceptT $ makeHttpRequest uri HEAD 0.3
```

Comment:
> I think we could use tryE instead of liftIO . runExceptT here

hm... I can see we have `transformers` in our dependencies, but in my generated `cabal.project.freeze` there are no
I have only 
```
any.exception-transformers ==0.4.0.9,
any.data-accessor-transformers ==0.2.1.7,
any.transformers-base ==0.4.5.2,
any.transformers-bifunctors ==0.1,
transformers-compat +five-three,
any.transformers-compat ==0.6.6,
any.transformers-fix ==1.0,
any.writer-cps-transformers ==0.5.6.1,
```

Reply:
I'll use `catchError`